Docs 5784 4.2 secure configuration properties kt (#796)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/secure-configuration-properties.adoc

include::_attributes.adoc[]
endif::[]

> 

Reviewer: Outline for Organizational Edit Suggestions:

SECURE CONFIGURATION PROPERTIES (adj. or verb?)

The text under this heading should introduce this topic as a whole and tell WHY your audience wants or needs to secure/to create secure config. properties. 
It should also prepare the audience for the first-level content by mentioning attributes, creating and defining, encrypting, config with module, file-level encryption, use of policies, supported algorithms and modes, FAQs, and additional resources, in that order (the order in which they subsequently appear).

==  Attributes
==  Create a Secure Configuration Properties File and Define Its Properties

The text under this heading should introduce the file and its properties and why your audience wants or needs to know about them. 
It should be followed by a subtopic Create a Secure Configuration Properties File, to be a peer to the next subtopic.
If you have just one subtopic under a topic, you should either merge that subtopic into the main topic or fine another subtopic. You need at least two subtopics to subdivide.

===   Define Secure Configuration Properties
==  Encrypt and Decrypt File Values with the Secure Properties Tool

Because this title mentions "file values" but is followed by "values" and then "files," I have two questions: are files and values separate things? Can you describe them in the same order as they are introduced?
One other thing: choose between "and" or "or" between "encrypt" and "decrypt," and use that only instead of both.

===   Encrypt or Decrypt Single Values
===   Encrypt or Decrypt Files
==  Configure the File with the Mule Secure Configuration Properties Module

Same as the previous first-level topics, this one introduces this task using the module and mentions the two next-level topics: Secure in XML and Securie in Studio.

===   Configure Secure Configuration Properties in XML

This text introduces the next two and tells why/how they are different.

====    Secure Configuration Properties with Default Values
====    Secure Configuration Properties with Custom Values
===   Configure Secure Configuration Properties in Studio

Ditto: introduce and list.

====    Install the Extension Module
====    Add and Configure Secure Configuration Properties to your App
====    Verify Secure Configuration Properties at Design Time
==  File Level Encryption
==  Usage of Secure Configuration Properties

This introduces encryption and nonencryption and why audience cares, how it chooses.

====    Use of Encrypted Secure Properties
====    Use of Non-encrypted Properties
===   Set Secure Configuration Properties File Dynamically

This introduces dynamic config of config props vs global config props.

====    Secure Configuration Properties with Environment File Defined
====    Secure Global Configuration Properties with Environment File Defined
===   Work with More than One Configuration File

Note that is a solitary subtopic, so it should be merged into the topic before it or a second peer topic found.

==  Supported Algorithms and Modes
===   Supported Algorithms
===   Supported Modes
==  Frequently Asked Questions
==  See Also

Finally, take a very BROAD view and see if all the first-level topics are really peers, then all the second-level, and the third-level, just to be sure the chronology of The User Journey is clear.  :-)

> To create secure configuration properties by encrypting them, review the following process:

Reviewer: See previous suggestion for the breadth of content coverage that this topic-set introduction should have.
This text introduces only the main tasks, but this set contains much more, including an attribute reference, file-level encryption, secure property use, algorithms and modes, a FAQ, and links to more info.

. Create a secure properties file and define its properties.

> . Encrypt and decrypt the file values.

Reviewer 2: Decrypt?

Author: Updated.

> . Encrypt and Decrypt the file values.

Reviewer: ```suggestion
. Encrypt and decrypt the file values.
```

Author: Updated.

> . Configure the secure properties file with the Mule Secure Configuration Properties Module.

Reviewer: I looked up this name online and got a couple of variations to Mule Secure Configuration Properties Module:

     the Mule Secure Configuration Property Extension module
     the Secure Configuration Properties Module
     Mule Secure Configuration Property Extension

What is the official name?
Note that a proper noun (all words init. capped) does not take an article ("a" or "the").

Author: Mule Secure Configuration Properties Extension (module)

== Attributes

> First identify the secure configuration properties attributes and encrypting attributes used in the process:

Reviewer: First = "Before you can <perform which of the actions we have identified so far?>....

Reviewer: What do we mean by "identify"? "Set," as in "configure"? Or "discover" or learn some values already established? Or some other meaning?

Author: Updated.

[cols="1,3", options="header"]
|===

> | Secure Configuration Properties Attributes |

Reviewer: Consider a list format rather than a table:

Properties attributes are:
     o Name
        A unique....
     o Key
        The word....
.
.
.
The properties encryption attributes are:

<same>

> | The word or phrase to unlock the properties value according to the system property that you define in this field. For example, `${production.myproperty}` instructs the Mule runtime to demand the key at runtime.

Reviewer: ```suggestion
| A word or phrase that you specify to unlock the properties value. For example, `${production.myproperty}` instructs the Mule runtime engine to demand the key at runtime.
```

Reviewer: ...but I don't understand how "production.myproperty" obviously relates to demanding a key.
This is probably my lack of technical understanding, which your audience will have more of than I do, but, to me, the words "production" and "myproperty" do not automatically imply "runtime" or "key."

> | The procedure that allows the Mule runtime to repeatedly use a block cipher with a single key. See <<supported_modes>> for a full list.

Reviewer: GLOBALLY change "Mule runtime" to either "Mule runtime engine" or "Mule" (after establishing "Mule" as the official short form of "Mule runtime engine").

Reviewer: Here and earlier, consider moving cross-reference links to See Also section. GLOBAL

|===

> == Create a Secure Configuration Properties File and Define Its Properties

Reviewer: Is this two tasks or one?
See org. edit comments.

> You can create a `.yaml` configuration file or a Spring-formatted `.properties` file. The recommended approach is to use a YAML configuration file, because it allows the addition of type validations and autocompletion. The Secure Configuration Properties module allows you to configure these supported file types. +

Reviewer: To what purpose is our audience creating these files? WHY do they want to do this?

Reviewer: Are "these supported file types" the YAML file and the .properties file?

Reviewer: I think the info in this paragraph might not be in the best order. Is the point that even though the module enables you to create either ".yaml" or ".properties," you want to use the module to create .yaml so that you can benefit from type validation and autocompletion? 
And how does this relate to the two topics contained within: configuring content and defining properties? Are these two tasks or just two steps of one task?

Reviewer: Hmm... How about something like...:

The first task in securing configuration properties is to create either a .yaml configuration file or a Spring-formatted .properties file. The Secure Configuraton Properties module enables you to create either file type, but using .yaml has the advantage of supporting type validation and autocompletion.

You create your properties file either in `src/main/resources` in your Mule project or by using absolute paths. The file can include both secured and nonsecured properties, and it can contain both encrypted and nonencrypted values. You encrypt values using `![value]`.

> Create these files under `src/main/resources` inside your Mule project, or you can also use absolute paths. The secure properties configuration file can include secured and non-secured properties.

Reviewer: Are "these files" both the .yaml and .properties? If so, why are we describing both if we recommend only one? 
See suggested rewording, which does not use subtopics and merges the info currently in the subtopics.
This rewording might not be technically accurate, but perhaps it provides an idea of a way to clarify the first task of the three?

> === Define Secure Configuration Properties

Reviewer: See above

Define secure properties in the file by encrypting the values between the sequence `![value]`. The file can contain encrypted and non-encrypted values that could be used anywhere in the Mule application.  +

> The following `file1.yaml` example contains both encrypted and non-encrypted configuration properties values:

Reviewer: nonencrypted

...unless you believe that it is confusing without the hyphen. See the entry for "prefixes" in the style ref.

[NOTE]

> The encrypted value must be enclosed in quotes so that it is read as a String.

Reviewer: string
(lowercase)

> The following `file1.properties` example contains the same previous configuration properties but uses a Spring-formatted properties file:

Reviewer: ```suggestion
The following `file1.properties` example contains the configuration properties used in the previous example but uses a Spring-formatted properties file:
```

testPropertyB=testValueB
----

== Encrypt and Decrypt File Values with the Secure Properties Tool

> You can download the https://docs.mulesoft.com/downloads/mule-runtime/4.2/secure-properties-tool.jar[JAR file] to use the Secure Properties Tool in the command line. The tool allows you to encrypt or decrypt single values, and complete files (both `.yaml` and `.properties` files). The tool supports three encryption modes: `string`, `file`, and `file-level`. +

Reviewer: ```suggestion
You can download the https://docs.mulesoft.com/downloads/mule-runtime/4.2/secure-properties-tool.jar[JAR file] to use the secure properties tool in the command line. The tool enables you to encrypt or decrypt both single values and complete files (both `.yaml` and `.properties` files). The tool supports three encryption modes: `string`, `file`, and `file-level`. +
```

> Note that key sizes might depend on the Java Cryptography Extension (JCE) version. Since version 1.8.0_161, the AES key size needs to be of at least 32 characters (256 bytes), instead of 16 characters in previous versions.

Reviewer: ```suggestion
Note that beginning with Jave Cryptography Extension (JCE) 1.8.0_161, the AES key size must be at least 32 characters (256 bytes).
```

[WARNING]

> When using encrypted properties, it is especially important to **secure access to the operating system**. Anyone who can run a `ps` command or view a Java console will be able to see the decrypted values that are stored in the Mule app's memory.

Reviewer: ```suggestion
When you use encrypted properties, it is especially important that you secure access to the operating system. Anyone who can run a `ps` command or view a Java console can see the decrypted values stored in the Mule app's memory.
```

> === Encrypt or Decrypt Single Values

Reviewer: As noted in the Org. edit, should we include information about when the audience should do this for single values vs. complete files?

> Use the `string` mode to encrypt or decrypt a single value provided as a parameter. Run the following command:

Reviewer: Could this and the following text be replaced with...:

Using `string` mode, encrypt or decrypt a single value provided as a parameter by replacing that value with any string you want to process:

[source,console.linenums]
java -jar secure-properties-tool.jar string encrypt Blowfish CBC mulesoft "some value to encrypt"

[source,console.linenums]
----
java -jar secure-properties-tool.jar string <encrypt|decrypt> <algorithm> <mode> <key> <value>
----

Replace `<value>` with any desired string value you want to process, for example:

[source,console.linenums]
----
java -jar secure-properties-tool.jar string encrypt Blowfish CBC mulesoft "some value to encrypt"
----

=== Encrypt or Decrypt Files

> The encryption tool is able to encrypt complete `.yaml` or `.properties` files.

Reviewer: You can use the secure properties tool to encrypt both .yaml and .properties files.

> * Use the `file` mode to encrypt all values inside a `.properties` file. The output is a file with the same properties, but its values are encrypted. Run the following command:

Reviewer: See edit to similar text in the single value encryption text.

Reviewer: QUESTION:
We state specifically that we recommend using .yaml rather than .properties. So...do we really want to use .properties examples or even mention using that file type after that statement? Are we documenting one "golden" path or multiple options?

+
[source,console.linenums]
----
java -jar secure-properties-tool.jar file <encrypt|decrypt> <algorithm> <mode> <key> <input file> <output file>
----
+
For example, assume that this is the input file:
+
`example_in.yaml`
+
[source,yaml,linenums]
----
properties:
  example1:
    value: "Muleman is here"
  example2: "Max the mule"
----
+

> If you run the command:

Reviewer: Run the following command:

(GLOBALLY avoid splitting sentences around examples. This might not be the most elegant solution; feel free to come up with your own.)

+
----
java -jar secure-properties-tool.jar file encrypt Blowfish CBC mulesoft example_in.yaml example_out.yaml
----
+
The output file is:
+
`example_out.yaml`
+
[source,yaml,linenums]
----
properties:
  example1:
    value: "![qCReIPK3jcqD7WR84ISSIQ==]"
  example2: "![En8lII21ZHrdIaINw0+mSA==]"
----

* Use the `file-level` mode to completely encrypt the properties file. The output is a file that is entirely encrypted.

> Using as input the `example_in.yaml` file from the previous example, if you run the command:

Reviewer: ```suggestion
Using as input the `example_in.yaml` file from the previous example, run the following command:
```

----
java -jar secure-properties-tool.jar file-level encrypt Blowfish CBC mulesoft example_in.yaml example_out.yaml
----

The output file is: +
`example_out.yaml`

[source,txt,linenums]
----
k/J1EHyxJagOcyQELoaqIDCzzKc1bLGJiHY2dv6jzuhtVpP1LTlgK3y8EH7OkK5iuuu8kik3P5dCDMixZXwn/EV6xbUwLFKnEb4x1pUX6aopLP/KULm0pQ==
----

> == Configure the File with the Mule Secure Configuration Properties Module

Reviewer: As noted earlier, verify what the exact official name is.

You can configure a secure properties file both in standalone XML or in Anypoint Studio, with the Mule Secure Configuration Properties Extension module available from https://anypoint.mulesoft.com/exchange/com.mulesoft.modules/mule-secure-configuration-property-module/[Exchange].

=== Configure Secure Configuration Properties in XML

> In your XML flow, use the Secure Configuration Properties module `<secure-properties:config>` to configure the file. Review the following example: +

Reviewer: ```suggestion
In your XML flow, use the Secure Configuration Properties module `<secure-properties:config>` to configure the file: +
```

> *Secure Configuration Properties with Default Values*

Reviewer: omit
Avoid giving any graphics or tables their own headings. Use a text intro.

[source,xml,linenums]
----
<secure-properties:config key="${runtime.property}"
  file="file1.yaml" name="test">
    <secure-properties:encrypt/>
</secure-properties:config>

<global-property name="prop" value="my-${secure::property.key1}"/>
----

> In the previous example: +

Reviewer: Consider omitting. The fact that this list directly follows the graphic makes the connection clear.

* The `<secure-properties:encrypt>` tag is required even when using the default values.
* The prefix `secure::` is required for loading the key.

> [NOTE]

Reviewer: This could be plain text instead of a note. Your choice, but know that some readers give notations far too much attention, while others skip them altogether. They tend to interrupt narrative flow. GLOBAL

> You have to use the `secure::` prefix to access all values inside a secure properties file, even if the values are non-encrypted. This behavior allows you to switch between files (or environments) without changing the entire configuration.

Reviewer: ```suggestion
You have to use the `secure::` prefix to access all values inside a secure properties file, even if the values are not encrypted. Doing this enables you to switch between files (or environments) without changing the entire configuration.
```

> * A decryption key is passed into the Mule runtime engine at deployment time as a system environment variable `runtime.property`. This property must be the exact key used to encrypt the values stored in the `file1.yaml` file.

Reviewer: ```suggestion
* At deployment, a decryption key is passed into the Mule runtime engine as a system environment variable `runtime.property`. This property must be the exact key used to encrypt the values stored in the `file1.yaml` file.
```

* The default encryption algorithm and mode is used.

> * If the actual (decrypted) value of the property `property.key1` is, for example, `"property"`, the value of the property `prop` will be `"my-property"`.

Reviewer: will be > is

> You can also change the encryption algorithm and crypto mode by setting additional attributes of the `<secure-properties:encrypt>` tag, for example:

Reviewer: ```suggestion
You can also change the encryption algorithm and crypto mode by setting additional attributes of the `<secure-properties:encrypt>` tag:
```

> *Secure Configuration Properties with Custom Values*

Reviewer: ditto remove heading

[source,xml,linenums]
----
<secure-properties:config key="${runtime.property}"
  file="file1.properties" name="test">
    <secure-properties:encrypt algorithm="AES" mode="CBC"/>
</secure-properties:config>
----

> In the previous example: +

Reviewer: ```suggestion
In the previous example, the `<secure-properties:encrypt>` tag uses the `AES` algorithm with `CBC` mode.
```

> * The `<secure-properties:encrypt>` tag uses the `AES` algorithm with `CBC` mode.

Reviewer: omit
A single item should not be a list.

=== Configure Secure Configuration Properties in Studio

> In Anypoint Studio, you can use this extension module by adding it as a dependency in your Mule app.

Reviewer: Advantage? Anything else to say about why your audience wants or needs to do this?

==== Install the Extension Module

. Open your Mule project in Anypoint Studio.

> . Go to the Mule Palette.

Reviewer: Unless there is a label in the UI that says "Mule Palette," lowercase "palette."

> . Select *Search in Exchange*, and search for the *Mule Secure Configuration Property Extension*.

Reviewer: Omit "the."

> . Add the extension.

Reviewer: How? What action does the user take to "add": drag, click, other?

==== Add and Configure Secure Configuration Properties to your App

. Go to your Mule app configuration file.

> . Select the *Global Elements* tab.

Reviewer: Unless the UI is crowded, avoid identifying "the tab" by anything other than label: Select *Global Elements*.

> . Click the *Create* button.

Reviewer: Ditto: Click *Create*.

. From the search bar, select *Secure Properties Config*.

> . Configure the global element with a *File* location, *Key*, *Algorithm*, *Mode*, *File level encryption*, and *Encoding*.

Reviewer: Do we have to name all of the UI elements or can we say "<Complete the text boxes> <or Click/Select values> to configure your global element"?

. Click *OK*.

image::secure-configuration-properties-studio.png[config extension]

==== Verify Secure Configuration Properties at Design Time
At design time in Studio, you can verify that secure configurations properties for a connector are set up correctly by testing the connector connection.

> The following example shows a Salesforce connector configuration where its secure properties (username, password and token) are encrypted in a `.yaml` file. +

Reviewer: ```suggestion
The following example shows a Salesforce connector configuration with its secure properties (username, password, and token) encrypted in a .yaml file. +
```

Reviewer: Consider ending the previous sentence with a colon and moving the screen capture here, following it with the text about the secure prefix.

> The `secure::` prefix is added before the property name definition, in order to access all values inside a secure properties file, even if the values are non-encrypted.

Reviewer: ```suggestion
The `secure::` prefix is added before the property name definition, to enable access to all values inside a secure properties file, even if the values are not encrypted.
```

image::secure-configuration-properties-studio2.png[config extension2]

> Click *Test Connection* to test the functionality of the connection and assure that the secure properties are read successfully.

Reviewer: This sounds like it should be a step in a task, so consider rewording to something like, "You can click ...to verify that the connection is functioning as expected."

> == File Level Encryption

Reviewer: File-Level Encryption

> Since version 1.1.0 of the Secure Configuration Properties module, you can encrypt the entire file (not only the values of the properties). To use a file that its encrypted:

Reviewer: ```suggestion
Beginning with Secure Configuration Properties module 1.1.0, you can encrypt an entire file rather than only individual proper values.
```

Reviewer: Add to this a statement about WHY this is a good thing and when your audience might want or need to do this instead of encrypting at the property level.

Reviewer: Follow this with another statement about how you can do this either in the XML flow (command line?) or in Anypoint Studio.

> In your XML flow, you add the `fileLevelEncryption` attribute and set it to `true`:

Reviewer: Omit "you"

*Secure Configuration Properties with Custom Values*
[source,xml,linenums]
----
<secure-properties:config key="${runtime.property}"
  file="file1.properties" name="test">
    <secure-properties:encrypt algorithm="AES" mode="CBC"/>
</secure-properties:config>
----
In the previous example: +

* The `<secure-properties:encrypt>` tag uses the `AES` algorithm with `CBC` mode.

=== Configure Secure Configuration Properties in Studio

In Anypoint Studio, you can use this extension module by adding it as a dependency in your Mule app.

==== Install the Extension Module

. Open your Mule project in Anypoint Studio.

. Go to the Mule Palette.

. Select *Search in Exchange*, and search for the *Mule Secure Configuration Property Extension*.

. Add the extension.

==== Add and Configure Secure Configuration Properties to your App

. Go to your Mule app configuration file.

. Select the *Global Elements* tab.

. Click the *Create* button.

. From the search bar, select *Secure Properties Config*.

. Configure the global element with a *File* location, *Key*, *Algorithm*, *Mode*, *File level encryption*, and *Encoding*.

. Click *OK*.

[NOTE]

> For *Encoding* the value *- -Empty- -* corresponds to the default value of UTF-8.

Reviewer 2: THis is stated in the default value for Encoding attribute

Author: Removed.

image::secure-configuration-properties-studio.png[config extension]

[NOTE]

> Use the `secure::` prefix to access all values inside a secure properties file, even if the values are non-encrypted. This behavior allows you to switch between files (or environments) without changing the entire configuration. +

Reviewer 2: This was already stated above

Author: Removed double note, and added the comment under the example explanation. Some users might only read the Studio section and not read the XML version. Therefore, the information should be indicated for both scenarios.

</secure-properties:config>
----

> In Studio, in the Secure Configuration Properties window, select *True* for the *File level encryption*:

Reviewer: Omit "the"

image::secure-configuration-properties-studio3.png[config extension3]

> == Usage of Secure Configuration Properties

Reviewer: Maybe these example topics should reside under a heading of Secure Configuration Property Use Cases?
If so, be sure to address the heading hierarchy, so that the first examples have their own subtopic title just like the Dynamic and More Than One titles....

The following example shows how a flow uses the encrypted `encrypted.value1` value to set the payload. At runtime, the decryption algorithm is used to store the decrypted value of `encrypted.value1` into memory.

> *Use of Encrypted Secure Properties*

Reviewer: ditto omit heading

[source,xml,linenums]
----
<flow name="main">
    <set-payload value="${secure::encrypted.value1}"/>
</flow>
----

> The following example shows how flow uses a non-encrypted value from the same secure properties file:

Reviewer: a flow, the flow

nonencrypted (GLOBAL if you agree)

> *Use of Non-encrypted Properties*

Reviewer: ditto

[source,xml,linenums]
----
<flow name="mainNonEncrypted">
    <set-payload value="${secure::testPropertyA}"/>
</flow>
----

=== Set Secure Configuration Properties File Dynamically

> A common configuration use case is to set the file to depend on a property (for example, `env`) to determine which file to use. For example, to use a `development-properties` file in development stage (which could not be encrypted) or a production file (with encrypted values).

Reviewer: ```suggestion
A common configuration use case is to set the file to depend on a property (for example, `env`) to determine which file to use: for example, to use a `development-properties` file during the development stage (which could not be encrypted) or a production file (with encrypted values).
```

> *Secure Configuration Properties with Environment File Defined*

Reviewer: ditto GLOBAL re. titles for graphics, examples, tables, lists, and so on.
Use intro text that ends with a colon instead.
I'll stop marking.

[source,xml,linenums]
----
<secure-properties:config key="${runtime.property}" file="${env}-properties.yaml" name="test"/>
----

> This way, the value of the property `env` determines which file to use to load the secure configuration properties. That `env` property could be set by a `global property`, system property, or environment property (see xref:configuring-properties.adoc[Configure Properties]).

Reviewer: ```suggestion
Using this method, the value of the property `env` determines which file to use to load the secure configuration properties. That `env` property could be set by a `global property`, system property, or environment property (see xref:configuring-properties.adoc[Configure Properties]).
```

*Secure Global Configuration Properties with Environment File Defined*
[source,xml,linenums]
----
<global-property name="env" value="dev"/>

<secure-properties:config key="${runtime.property}" file="${env}-properties.yaml" name="test"/>
----

> This way, the default value for the `env` property will be `"dev"`, which can still be overriden with a system or environment property. Please note that this is *required* for metadata resolution in Anypoint Studio. If you do not define default values for the properties that are passed through the command line you will receive an error while creating application model for all message processors that depend on them.

Reviewer: ```suggestion
Using this method, the default value for the `env` property is `"dev"`, which can still be overridden with a system or environment property. Note that this is required for metadata resolution in Anypoint Studio. If you do not define default values for the properties that are passed through the command line, you receive an error when you create the application model for all message processors that depend on those properties.
```

> === Work with More than One Configuration File

Reviewer: ```suggestion
=== Work with More Than One Configuration File
```

> You can define more than one secure configuration properties file to read properties from. To do so, simply define a `<secure-properties:config />` tag for each file you want to load. Each secure configuration properties file is independently configured with an encryption algorithm, cipher mode, and encryption/decryption key. None of these settings needs to be the same between secure configuration properties files. +

Reviewer: ```suggestion
You can define more than one secure configuration properties file. To do so, simply define a `<secure-properties:config />` tag for each file you want to load. Each secure configuration properties file is independently configured with an encryption algorithm, cipher mode, and encryption/decryption key. These settings do not have to be the same among the files. +
```

* ECB
* OFB

> == Frequently Asked Questions

Reviewer: Consider adding some text here that ties this topic into the set and introduces how the audience should use it.
Maybe something like...:

"After you learn about, create, and use secure configuration properties files for a while, you might have additional questions about troubleshooting and maximizing their use. Following are some common questions and answers. Refer as well to the topics linked to in See Also.

*What if a property is defined in multiple files?*

> In that case, the actual property's value will be the one in which is first defined.

Reviewer: ```suggestion
If a property is defined differently in more than one file, its actual value is the one first defined.
```

Reviewer: Do we need to say anything that explains how "first defined" is determined or what that is based on?

> *What if I want to define a secure property to depend on a previously defined secure property?*

Reviewer: ```suggestion
*Can I define a secure property that depends on a previously defined secure property?*
```

Reviewer: Or do we mean, "Can I create a new secure property that is based on a previous one?" or "...that depends on a previous one?"

> It is possible to define a property to depend on a previously defined one, just using the same syntax: `property=My dependent ${secure::dependent.property} value`.

Reviewer: ```suggestion
Yes. Just use the same syntax: `property=My dependent ${secure::dependent.property} value`.
```

> Note that Secure Configuration Properties can depend on other Secure Configuration Properties, but not on other Mule configuration properties.

Reviewer: ```suggestion
Note that secure configuration properties can depend on other secure configuration properties, but not on other Mule configuration properties.
```

Reviewer: Do we need a "For example, ..."?
